### PR TITLE
Speedup stackitem json serialization

### PR DESCRIPTION
**Pull request by fyrchik**
Cache already marshaled values.
The expected memory overhead is linear (map itself + living references to the old buffer memory in case it grows with reallocation).

### CODE REVIEW

**Comment by roman-khimov** [comment]
> living references to the old buffer memory in case it grows with reallocation

`map[Item]struct{start int, end int}`?

File: pkg/io/binaryWriter.go [hunk 1]
```diff
@@ -106,23 +106,30 @@ func (w *BinWriter) WriteVarUint(val uint64) {
...
+		data = append(data, byte(val), byte(val>>8), byte(val>>16), byte(val>>24))
+		return data
 	}
 
-	w.WriteB(byte(0xff))
-	w.WriteU64LE(val)
+	data = append(data, byte(0xff))
+	data = append(data,
+		byte(val), byte(val>>8), byte(val>>16), byte(val>>24),
+		byte(val>>32), byte(val>>40), byte(val>>48), byte(val>>56))
```
Comment (roman-khimov) [inline pkg/io/binaryWriter.go:131]: It'd be interesting to compare this with `encoding/binary` in terms of raw speed.

File: pkg/vm/stackitem/serialization.go [hunk 2]
```diff
@@ -19,38 +19,41 @@ var ErrUnserializable = errors.New("unserializable")
 
 // serContext is an internal serialization context.
 type serContext struct {
-	*io.BinWriter
-	buf          *io.BufBinWriter
+	data         []byte
+	protected    bool
```
Comment (roman-khimov) [inline pkg/vm/stackitem/serialization.go:23]: I don't think we need this `protected` var, just always enforce size limit.

File: pkg/vm/stackitem/serialization.go [hunk 3]
```diff
@@ -19,38 +19,41 @@ var ErrUnserializable = errors.New("unserializable")
...
 // interface.
 func EncodeBinary(item Item, w *io.BinWriter) {
 	sc := serContext{
-		BinWriter:    w,
+		data:         nil,
 		allowInvalid: false,
 		seen:         make(map[Item]bool),
 	}
-	sc.serialize(item)
+	err := sc.serialize(item)
```
Comment (roman-khimov) [inline pkg/vm/stackitem/serialization.go:51]: So this effectively could be replaced with `data, err := Serialize(item)`.

File: pkg/vm/stackitem/serialization.go [hunk 4]
```diff
@@ -59,89 +62,100 @@ func EncodeBinary(item Item, w *io.BinWriter) {
 // (like recursive array) is encountered it just writes special InvalidT
 // type of element to w.
 func EncodeBinaryProtected(item Item, w *io.BinWriter) {
-	bw := io.NewBufBinWriter()
 	sc := serContext{
-		BinWriter:    bw.BinWriter,
-		buf:          bw,
+		protected:    true,
```
Comment (roman-khimov) [inline pkg/vm/stackitem/serialization.go:66]: While this still differs with `allowInvalid`.

File: pkg/vm/stackitem/serialization.go [hunk 5]
```diff
@@ -76,10 +76,18 @@ func EncodeBinaryProtected(item Item, w *io.BinWriter) {
 }
 
 func (w *serContext) serialize(item Item) error {
-	if w.seen[item] {
-		return ErrRecursive
+	if v, ok := w.seen[item]; ok {
+		if v.start == v.end {
```
Comment (roman-khimov) [inline pkg/vm/stackitem/serialization.go:80]: Nice trick.

**Review by roman-khimov — COMMENTED** [review]
I'm a little worried about our test coverage after these changes. Do we hit every corner case with our current test set? But it looks good otherwise.

File: pkg/vm/stackitem/serialization.go [hunk 6]
```diff
@@ -118,9 +126,9 @@ func (w *serContext) serialize(item Item) error {
 				return err
 			}
 		}
-		delete(w.seen, item)
+		w.seen[item] = sliceNoPointer{start, len(w.data)}
```
Comment (AnnaShaleva) [inline pkg/vm/stackitem/serialization.go:129]: Don't we want to move this line out of switch (i.e. cache other stack item types also, like it is done for JSONisation)?

File: pkg/io/binaryWriter.go (hunk 1 above)
Comment (fyrchik) [reply pkg/io/binaryWriter.go:131]: I have compared this with old implementation (via `binary.LittleEndian`, but with `WriteU64LE` wrappers removed).
The difference in performance is negligible (big uint64 number):
```
BenchmarkVarint/append-8         	308960446	         3.636 ns/op
BenchmarkVarint/raw-8            	361104601	         3.209 ns/op
```

File: pkg/vm/stackitem/serialization.go (hunk 6 above)
Comment (fyrchik) [reply pkg/vm/stackitem/serialization.go:129]: I have checked this now:
```
func BenchmarkEncodeBinaryWithDup(b *testing.B) {
	bi := NewBigInteger(big.NewInt(123))
	arr := NewArray([]Item{
		NewBigInteger(big.NewInt(123)),
		bi,
		NewBool(true),
		bi,
		NewBool(false),
		bi,
	})

	benchmarkEncodeBinary(b, arr)
}

func BenchmarkEncodeBinaryCopy(b *testing.B) {
	arr := NewArray([]Item{
		NewBigInteger(big.NewInt(123)),
		NewBigInteger(big.NewInt(123)),
		NewBool(true),
		NewBigInteger(big.NewInt(123)),
		NewBool(false),
		NewBigInteger(big.NewInt(123)),
	})

	benchmarkEncodeBinary(b, arr)
}
```

Results:
```
name                   old time/op    new time/op    delta
EncodeBinaryWithDup-8     739ns ±14%     771ns ±10%     ~     (p=0.123 n=10+10)
EncodeBinaryCopy-8        743ns ±12%     886ns ± 6%  +19.33%  (p=0.000 n=10+10)

name                   old alloc/op   new alloc/op   delta
EncodeBinaryWithDup-8      480B ± 0%      448B ± 0%   -6.67%  (p=0.000 n=10+10)
EncodeBinaryCopy-8         480B ± 0%      480B ± 0%     ~     (all equal)

name                   old allocs/op  new allocs/op  delta
EncodeBinaryWithDup-8      10.0 ± 0%       8.0 ± 0%  -20.00%  (p=0.000 n=10+10)
EncodeBinaryCopy-8         10.0 ± 0%      10.0 ± 0%     ~     (all equal)
```

Comment (fyrchik) [reply pkg/vm/stackitem/serialization.go:129]: So I think it doesn't worth the effort (I don't expect a lot of shared references for primitive items).
But we can behave the same way in `ToJSON` then.

Comment (fyrchik) [reply pkg/vm/stackitem/serialization.go:129]: Fixed `ToJSON`.

**Review by AnnaShaleva — APPROVED** [review]

File: pkg/vm/stackitem/serialization.go [hunk 7]
```diff
@@ -68,6 +68,7 @@ func EncodeBinaryProtected(item Item, w *io.BinWriter) {
 	}
 	sc.serialize(item)
 	if bw.Err != nil {
+		fmt.Println(bw.Err)
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/vm/stackitem/serialization.go]: Still there in the first commit.

File: pkg/io/binaryWriter.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/io/binaryWriter.go:131]: So `raw` is `encoding/binary`?

Comment (fyrchik) [reply pkg/io/binaryWriter.go:131]: Yes

Comment (roman-khimov) [reply pkg/io/binaryWriter.go:131]: 13%, not nothing. What if we `append` 8 bytes and then `PutUint64` into them?

Comment (fyrchik) [reply pkg/io/binaryWriter.go:131]: Fixed.
